File Expanding Utility for Nested Code Files

jim-hart opened the pull request:
This serves as an alternative to Transcrypts `-xpath` option, primarily for those who have problems with getting Transcrypt to recognize nested modules.

`file_expander.py` is composed of a single class, `FileExpander`, that will search for user-created screeps code under `screeps-starter-python/src/`.  The files can be directly in `src/`, or in any type of nested directory structure within `src/`.  By providing `-e` or `--expand-files` to `build.py`, before Transcrypt processes anything, all of the code files meeting the above description are copied to a `__py_build__` folder, which is located in `src/`.  The copy process flattens any nested folder structure created by the user and stores all files directly under `__py_build__`.  The user's original files remain untouched, and only updated files are actually copied.  

The following actions are automatically handled:

- Creation of the `__py_build__` folder if missing
- Copying of required folders (and their contents), like `defs/` if not found in `__py_build__`
- Automatic updating of `defs/` files and subdirectories in `__py_build__`: this ensures the file is updated in `__py_build__` if a newer version is released through `screeps-starter-python`, or if the user wishes to make custom changes
 
The build process, in my experience, isn't affected by this module.  `filecmp` caches previous comparisons for improved performance, and `shutil` is implemented so that it won't even attempt to copy a file if it hasn't been modified.

`build.py` defaults to the original build process, so those who don't want to use it don't have to make any changes to their workflow.  If desired, this can be used simultaneously with the standard build process as the `-e` switch directs which folder Transcrypt should target.  `__py_build__` can be deleted at any time without consequence.

daboross commented:
This looks good to me. I'd probably prefer reStructuredText / Sphinx styled doc comments for consistency, but that can always be fixed later, and it's a build script anyways.

If no one else has any concerns, I'll merge this shortly.

jim-hart commented:
@daboross - Docstring format for `file_expander.py` now matches the project's docstring format.  Also updated the format for the property I added to `Configuration`.

daboross commented:
Alright.

Thank you for doing this
